pkg/agent/filter.go

```diff
@@ -2,11 +2,55 @@ package agent
 
 import (
 	"fmt"
+	"net/netip"
 	"regexp"
 	"strings"
+
+	"golang.org/x/exp/slices"
 )
 
-type interfaceFilter struct {
+type InterfaceFilter interface {
+	Allowed(iface string) (bool, error)
+}
+
+type ipInterfaceFilter struct {
+	allowedIPs []netip.Prefix
+	// Almost always going to be a wrapper around getting
+	// the interface from net.InterfaceByName and then calling
+	// .Addrs() on the interface
+	ipsFromIface func(ifaceName string) ([]netip.Prefix, error)
+}
+
+func initIPInterfaceFilter(ips []string, ipsFromIface func(ifaceName string) ([]netip.Prefix, error)) (ipInterfaceFilter, error) {
+	ipIfaceFilter := ipInterfaceFilter{}
+	ipIfaceFilter.ipsFromIface = ipsFromIface
+
+	for _, ip := range ips {
```

Comment:
do we need special handling for loopback IPs or LinkLocal ?

Reply:
I don't think so, the code uses [net.InterfaceByName()](https://pkg.go.dev/net#InterfaceByName) and then calls [Interface.Addrs()](https://pkg.go.dev/net#Interface.Addrs) which is happy to return loopback or linklocal addresses.

Just running some code that loops through interfaces and calls `.Addrs()` on my machine:
```
if: lo
addrs: 127.0.0.1/8

if: lo
addrs: ::1/128

if: enp6s0
addrs: 10.0.0.100/24

if: enp6s0
addrs: fe80::b981:bd8e:91f6:b3e5/64
```